File: dotnet/asp.net.mvc5/Flex/Services/SecurityService.cs
/**
* Copyright (c) 2016 by CyberSource
* Governing licence: https://github.com/CyberSource/cybersource-flex-samples/blob/master/LICENSE.md
*/

using Flex.Models;
using Newtonsoft.Json;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.OpenSsl;
using Org.BouncyCastle.Security;
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Threading.Tasks;
using System.Web.Configuration;

namespace Flex.Services
{
    interface ISecurityService
    {
        Dictionary<string, string> GenerateSignature();
        Task<KeyResult> FetchKeystore();
        RsaKeyParameters DecodePublicKey(string derEncodedKey);
        bool Verify(AsymmetricKeyParameter publicKey, string dataToVerify, string base64Signature);
    }

    public class SecurityService: ISecurityService
    {
        private static string flexKeysEndpoint = "https://testflex.cybersource.com/cybersource/flex/v1/keys";

        private string mid;
        private string profileId;
        private string cmmKey;
        private string organizationId;
        private string keyStoreFile;
        private string keyStorePassword;
        private string privateKeyPassword;
                
        public SecurityService()
        {
            // Read in flex configuration from web.config    
            mid = WebConfigurationManager.AppSettings["mid"];
            profileId = WebConfigurationManager.AppSettings["profileId"];
            cmmKey = WebConfigurationManager.AppSettings["cmmKey"];
            organizationId = WebConfigurationManager.AppSettings["organizationId"];
            keyStoreFile = WebConfigurationManager.AppSettings["keyStoreFile"];
            keyStorePassword = WebConfigurationManager.AppSettings["keyStorePassword"];
            privateKeyPassword = WebConfigurationManager.AppSettings["privateKeyPassword"];
        }

        private AsymmetricKeyParameter readPrivateKey()
        {
            var certificate = new X509Certificate2(keyStoreFile, keyStorePassword, X509KeyStorageFlags.Exportable | X509KeyStorageFlags.PersistKeySet);

            // Private Key
            RSACryptoServiceProvider rsa = (RSACryptoServiceProvider)certificate.PrivateKey;
            MemoryStream memoryStream = new MemoryStream();
            TextWriter streamWriter = new StreamWriter(memoryStream);
            PemWriter pemWriter = new PemWriter(streamWriter);
            AsymmetricCipherKeyPair keyPair = DotNetUtilities.GetKeyPair(rsa);

            return keyPair.Private;
        }

        public Dictionary<string,string> GenerateSignature()
        {
            try
            {
                // data used to generate signature
                var time = (long)(DateTime.UtcNow - new DateTime(1970, 1, 1)).TotalMilliseconds;
                var contentType = "application/json";
                var httpVerb = "post";

                // Preparing data string that will be used to derive the signature
                var dataToSign = new StringBuilder();
                dataToSign.Append("cmm-key=").Append(cmmKey).Append(',');
                dataToSign.Append("content-type=").Append(contentType).Append(',');
                dataToSign.Append("http-verb=").Append(httpVerb).Append(',');
                dataToSign.Append("organization-id=").Append(organizationId).Append(',');
                dataToSign.Append("time=").Append(time);
                var dataToSignBytes = Encoding.UTF8.GetBytes(dataToSign.ToString());
                                
                // Signing data
                var privateKey = readPrivateKey();                                            

                ISigner signer = SignerUtilities.GetSigner("SHA512withRSA");
                signer.Init(true, privateKey);
                signer.BlockUpdate(dataToSignBytes, 0, dataToSignBytes.Length);
                byte[] sigBytes = signer.GenerateSignature();
                var encoded = Convert.ToBase64String(sigBytes); // transform byte array to Base64 string

                // Return dictionary of header names & values
                return new Dictionary<string, string>()
                {
                    { "Authorization", "CWS CWSAccessKeyId:" + encoded },
                    { "HTTP-Verb", httpVerb },
                    { "Organization-Id", organizationId },
                    { "Time", time.ToString() },
                    { "CMM-Key", cmmKey },
                    { "Content-Type", contentType }
                };
            }
            catch (Exception e)
            {
                throw e;
            }
        }

        public async Task<KeyResult> FetchKeystore()
        {
            // prepare keys endpoint request payload
            var payload = new KeyParameters()
            {
                profileId = profileId,
                encryptionType = "WebCryptoAPI"
            };

            // prepare HTTP headers to make server2flex rest call
            var signature = GenerateSignature();            
            
            byte[] buffer = Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(payload));

            // TODO this should probably be rewritten to use the newer HttpClient from .net 4.5
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(flexKeysEndpoint);
            request.Method = signature["HTTP-Verb"];
            request.Headers.Add("X-MERCHANT-ID", mid);
            request.Headers.Add("Authorization", signature["Authorization"]);
            request.Headers.Add("HTTP-Verb", signature["HTTP-Verb"]);
            request.Headers.Add("Organization-Id", signature["Organization-Id"]);
            request.Headers.Add("Time", signature["Time"]);
            request.Headers.Add("CMM-Key", signature["CMM-Key"]);
            request.ContentType = signature["Content-Type"];
            request.ContentLength = buffer.Length;

            var reqStream = request.GetRequestStream();
            reqStream.Write(buffer, 0, buffer.Length);
            reqStream.Close();            

            using (var response = (HttpWebResponse)request.GetResponse())
            {
                var encoding = Encoding.UTF8;

                using (var responseStream = response.GetResponseStream())
                using (var reader = new StreamReader(responseStream, encoding))
                {
                    var result = reader.ReadToEnd();
                    return await Task.Factory.StartNew(() =>
                    {
                        return JsonConvert.DeserializeObject<KeyResult>(result);
                    });
                }                    
            }            
        }

        public RsaKeyParameters DecodePublicKey(string derEncodedKey)
        {
            try
            {
                byte[] keyBytes = Convert.FromBase64String(derEncodedKey);
                return (RsaKeyParameters)PublicKeyFactory.CreateKey(keyBytes);
            }
            catch (Exception e)
            {
                throw e;
            }
        }

        public bool Verify(AsymmetricKeyParameter publicKey, string dataToVerify, string base64Signature)
        {
            try
            {
                var dataToVerifyBytes = Encoding.UTF8.GetBytes(dataToVerify.ToString());

                ISigner signer = SignerUtilities.GetSigner("SHA512withRSA");
                signer.Init(false, publicKey);
                signer.BlockUpdate(dataToVerifyBytes, 0, dataToVerifyBytes.Length);
                var signature = Convert.FromBase64String(base64Signature);
                return signer.VerifySignature(signature);
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
    }
}

File: demo-sites/spring-boot/src/main/java/com/cybersource/flex/samples/SecurityService.java
/**
* Copyright (c) 2016 by CyberSource
* Governing licence: https://github.com/CyberSource/cybersource-flex-samples/blob/master/LICENSE.md
*/

package com.cybersource.flex.samples;

import java.security.KeyFactory;
import java.security.KeyStore;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Security;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.ClassPathResource;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Service;

@Service
public class SecurityService {

    private static final Base64.Encoder ENCODER = Base64.getEncoder();
    private static final Base64.Decoder DECODER = Base64.getDecoder();

    @Value("${cmmKey}")
    private String cmmKey;
    @Value("${organizationId}")
    private String organizationId;
    @Value("${keyStoreFile}")
    private String keyStoreFile;
    @Value("${keyStorePassword}")
    private String keyStorePassword;
    @Value("${privateKeyPassword}")
    private String privateKeyPassword;

    public SecurityService() {
        // Add Bouncy Castle JCE provider dynamically.
        // http://www.bouncycastle.org/wiki/display/JA1/Provider+Installation
        Security.addProvider(new BouncyCastleProvider());
        // Note: to run this example you *must* install the unlimited policy files in the JVM you are using.
    }

    /**
     * This method adds HTTP authentication headers.
     * 
     * @param headers map to be added to
     */
    public void addSignature(HttpHeaders headers) {
        try {
            // data used to generate signature
            final long time = System.currentTimeMillis(); // current timestamp
            final String contentType = "application/json";
            final String httpVerb = "post";

            // Retrieving RSA private key from P12 keystore
            KeyStore keyStore = KeyStore.getInstance("PKCS12", BouncyCastleProvider.PROVIDER_NAME);
            ClassPathResource keyResource = new ClassPathResource(keyStoreFile);
            keyStore.load(keyResource.getInputStream(), keyStorePassword.toCharArray());
            PrivateKey privateKey = (PrivateKey) keyStore.getKey("serialNumber=" + cmmKey + ",CN=" + organizationId, privateKeyPassword.toCharArray());

            // Preparing data string that will be used to derive the signature
            StringBuilder dataToSign = new StringBuilder();
            dataToSign.append("cmm-key=").append(cmmKey).append(',');
            dataToSign.append("content-type=").append(contentType).append(',');
            dataToSign.append("http-verb=").append(httpVerb).append(',');
            dataToSign.append("organization-id=").append(organizationId).append(',');
            dataToSign.append("time=").append(time);

            // signing data
            Signature signature = Signature.getInstance("SHA512withRSA", BouncyCastleProvider.PROVIDER_NAME);
            signature.initSign(privateKey);
            signature.update(dataToSign.toString().getBytes());
            byte[] signed = signature.sign();
            String encoded = ENCODER.encodeToString(signed); // transform byte array to Base64 string

            // populating signature data to HTTP headers
            headers.set("Authorization", "CWS CWSAccessKeyId:" + encoded);
            headers.set("HTTP-Verb", httpVerb);
            headers.set("Organization-Id", organizationId);
            headers.set("Time", "" + time); // timestamp used to generate signature
            headers.set("CMM-Key", cmmKey);
            headers.set("Content-Type", contentType);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public PublicKey decodePublicKey(String derEncodedKey) {
        try {
            byte[] keyBytes = DECODER.decode(derEncodedKey);
            PublicKey publicKey = KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(keyBytes));
            return publicKey;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public boolean verify(final PublicKey publicKey, final String dataToVerify, final String base64Signature) {
        try {
            final Signature signInstance = Signature.getInstance("SHA512withRSA");
            signInstance.initVerify(publicKey);
            signInstance.update(dataToVerify.getBytes());
            byte[] signature = DECODER.decode(base64Signature);
            return signInstance.verify(signature);
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

}
